# Lab 1

## NLTK

Importo la librería de NLTK y descargo los textos disponibles en su paquete **NLTK book**

In [10]:
# !conda install nltk

In [1]:
import nltk

In [2]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

Importo el corpora descargado de la librería

In [11]:
from nltk.book import *

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


Listo 5 (los primeros) tokens del texto 1, Moby Dick by Herman Melville 1851

In [13]:
text1.tokens[0:5]

['[', 'Moby', 'Dick', 'by', 'Herman']

Listo 5 (los primeros) tokens del texto 7, Wall Street Journal

In [15]:
text7.tokens[0:5]

['Pierre', 'Vinken', ',', '61', 'years']

Con texts() listo todos los textos y utilizo count("palabra") para saber la cantidad de veces que se encuentra dicha palabra en el texto.

In [17]:
texts()

text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [18]:
text1.count("Moby")

84

## Strings y expresiones regulares

Para probar el funcionamiento de la librería formulo una oración, y con la función word_tokenize, extraigo en una lista todos los tokens del texto separados por espacios

In [20]:
mysent = "The cat sat on the mat."

In [21]:
from nltk import word_tokenize

In [22]:
mysent_tokens = word_tokenize(mysent)

In [23]:
mysent_tokens

['The', 'cat', 'sat', 'on', 'the', 'mat', '.']

Con len() cuento la cantidad de tokens en dicha lista (sin diferenciar con signos)

In [25]:
len(mysent_tokens)

7

### Expresiones regulares

Utilizando re.search, puedo usar una expresión regular para encontrar en el texto la coincidencia. En este caso "leng" en la oración "procesamiento del lenguaje natural".

In [28]:
from nltk import re

In [29]:
m = re.search("leng","procesamiento del lenguaje natural")

Puedo ver desde qué token a qué token se produce la coincidencia.

In [34]:
m

<re.Match object; span=(18, 22), match='leng'>

In [35]:
m.start()

18

In [36]:
m.end()

22

In [37]:
n = re.search("procesan", "procesamiento del lenguaje natural")

In [38]:
n

Usando una expresión regular para quitar, por ejemplo, los puntos, extraigo todos los tokens del texto, y por cada uno evalúo (con re.search) si es una palabra (\w) o no, en caso de serlo, la guardo en la lista.

In [47]:
mysent_tokens_nopunct = [word for word in word_tokenize(mysent) if re.search("\w", word)]

In [49]:
mysent_tokens_nopunct

['The', 'cat', 'sat', 'on', 'the', 'mat']

La función no reconoce igualdad entre una palabra empezada en mayúsculas y otra en minúsculas, por lo que coloco todas las letras en mayúsculas con upper()

In [50]:
mysent_tokens_nopunct_upper = [word.upper() for word in mysent_tokens_nopunct]

In [51]:
mysent_tokens_nopunct_upper

['THE', 'CAT', 'SAT', 'ON', 'THE', 'MAT']

In [52]:
len(mysent_tokens_nopunct)

6

Verificando la coincidencia con las mayúsculas, vemos que en un set (donde no se repiten los objetos), en este caso The y the, aparecen juntos sin haber usado el upper().

In [57]:
set(mysent_tokens_nopunct)

{'The', 'cat', 'mat', 'on', 'sat', 'the'}

In [58]:
mtnc_set = set(mysent_tokens_nopunct_upper)

In [59]:
mtnc_set

{'CAT', 'MAT', 'ON', 'SAT', 'THE'}

In [61]:
len(mtnc_set)

5

Ahora evalúo en un texto grande y no una oración, tomo los tokens del texto 1 (el de Moby Dick), quito los tokens de puntuación y listo los primeros 5

In [63]:
moby_dick_tokens = text1.tokens

In [64]:
moby_dick_tokens_nopunct = [word.lower() for word in moby_dick_tokens if re.search("\w", word)]

In [65]:
moby_dick_tokens_nopunct[0:5]

['moby', 'dick', 'by', 'herman', 'melville']

## Ejercicio

1. Cuál es el número de tokens en Moby Dick?

Para hacerlo tomo los tokens (.tokens), los coloco en minúscula (.loer) y de ellos sólo filtro las palabras (del for, solo filtro con re.search() las \w palabras)

In [103]:
tokens_moby_nopunct_lower = [word.lower() for word in text1.tokens if re.search("\w", word)]

In [104]:
len(tokens_moby_nopunct_lower)

218621

2. Cuál es el número de types en Moby Dick?

Busco la cantidad de palabras únicas pasando la lista a un set y contándolo

In [105]:
moby_dick_types = set(tokens_moby_nopunct_lower)

In [106]:
len(moby_dick_types)

17140

3. Moby Dick type-token ratio

Ahora calculo el type-token con los datos calculados: la cantidad de types y el total de tokens

In [132]:
type_token_ratio = len(moby_dick_types) / len(tokens_moby_nopunct_lower)

In [133]:
type_token_ratio

0.07840051962071347

4. WSJ type-token ratio

Uso la misma técnica de los puntos 1-2-3-4 para el text del Wall Street Journal

In [109]:
tokens_wsj_nopunct_lower = [word.lower() for word in text7.tokens if re.search("\w", word)]

In [110]:
wsj_types = set(tokens_wsj_nopunct_lower)

In [111]:
type_token_ratio2 = len(wsj_types) / len(tokens_wsj_nopunct_lower)

In [112]:
type_token_ratio2

0.129748424801388

5. Cuál de los dos tiene más diversidad léxica?

El texto del WSJ tiene más diversidad léxica, su type-token ratio es mayor

6. Puede pensar una razón por por la cual ese corpus es más diverso que el otro?

Al tratarse de noticias diarias que son espontáneas día a día (empresas nuevas, términos políticos nuevos, nombres nuevos) con lenguaje utilizado de manera corriente, la cantidad de palabras a utilizar que aparecerán una sola vez es mucha mas elevada que un libro que se rodea en un único ambiente literario.

7. Cual es el “Maximum Likelikhood Estimate (MLE)” de la palabra “whale” (ballena) en Moby Dick?

Para calcular el MLE de la palabra whale uso la fórmula unigram: P(palabra) = cantidad de veces que aparece / tokens totales

In [122]:
whale_pr = len([word for word in tokens_moby_nopunct_lower if re.search("whale", word)]) / len(tokens_moby_nopunct_lower)

In [123]:
whale_pr

0.007707402308103979

Hago lo mismo para el del WSJ

In [128]:
whale_pr = len([word for word in tokens_wsj_nopunct_lower if re.search("whale", word)]) / len(tokens_wsj_nopunct_lower)

In [129]:
whale_pr

0.0

Aplico smoothing como ejemplo para ver la diferencia del ratio nulo

In [130]:
whale_pr = (len([word for word in tokens_wsj_nopunct_lower if re.search("whale", word)])+1) / (len(tokens_wsj_nopunct_lower)+len(set(tokens_wsj_nopunct_lower)))

In [131]:
whale_pr

1.0103561505430664e-05